src/cmds/fs/sort.c

```diff
@@ -0,0 +1,166 @@
+/**
+ * @file
+ * @brief Sort lines of text files to standart output.
+ *
+ * @date 28.09.18
+ * @author Mark Kovalev
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <errno.h>
+
+static int blank, ign_case, rev;
+
+static void print_usage(void) {
+	printf(
+		"Usage: sort [OPTION]... [FILE]...\n"
+		"Write sorted lines of FILE to standard output.\n"
+		"\nOrdering options:\n\n"
+		"  -b\tignore leading blanks\n"
+		"  -f\tfold lower case to upper case characters\n"
+		"  -r\treverse the result of comparisons\n"
+		"\nOther options:\n\n"
+		"  -h\tdisplay this help and exit\n"
+		"\n*** WARNING ***\n"
+		"The locale specified by the environment affects sort order.\n"
+		"This implementation assumes that LC_ALL=C set to get the\n"
+		"traditional sort order that uses native byte values.\n"
+	);
+}
+
+static int compare(const void *fst_ptr, const void *snd_ptr) {
```

Comment:
It looks like a sort of string comparison. If it's so, can I ask to try `strcmp` instead?

Reply:
I need to compare strings with different conditions depending on flags used, which is not supported by `strcmp`. If it's needed though, I can use `strcmp` when there is no flags. But I'm not sure this will speed up comparision process.

Reply:
Yes, I understand, I think your solution is working well. But my proposition is concerning code simplification. Look, instead of `while (fst[i] == ' ')` you can use `while (isspace(fst[i])`, because the second variant prоcesses tabs and whitespaces, while the first one processes only white spaces.

Now, after skipping white spaces you have two options - usual string comparison, or comparison ignoring case (-f). So in case if option `-f` is specified, you can use `strncasecmp`, which compares two strings ignoring case. In case if `-f` was not specified, you can use usual `strcmp`.
What do you think about it? Probably, it can do this job?

Reply:
But in any case, it's just a proposition to think about it, not a requirement.

Reply:
Thanks for advice. I totally forgot about tabs, but they're need to be ignored with `-b` option. So I will use `isspace`. And I'll check if `strcasecmp` suits `-f` option, because it may be tricky how `sort` compares with case insensitivity.

Reply:
`isspace` works like a charm. `strcmp` suits my needs too. Problems appears with `strcasecmp`. It compares strings with lowering case of characters. But `sort -f` folds lower case to upper case. So sort results differs. If it's not vital to follow up GNU sort implemintation, I can leave things as it is. If needed, I can do GNU-like solution. What do you think?

Reply:
> It compares strings with lowering case of characters. But sort -f folds lower case to upper case. So sort results differs.

Hm, it's interesting.. Could you, please, paste here an example of two strings when `sort -f` and `strcasecmp` will get different results? I am not sure about all the languages in the world, but for English and many other European languages there should be no matter which case (lower or upper) is used during string comparision.. Are you using English language in your test files?

Reply:
By the way, since `sort` requires converting to upper case, could we firstly convert the string to upper case using `toupper` character by character, and then apply `strcmp` to the obtained strings?

Reply:
Yes, I'm using English language. I suppose, it's not language problem.
Here's input `atest` file:
```
Aab
AA
b
aa
AAb
Aa
B
aab
aA
aAb
```
Standard GNU sort ─ `LC_ALL=C sort -f atest`:
```
AA
Aa
aA
aa
AAb
Aab
aAb
aab
B
b
```
My sort with `strcasecmp` as string comparator ─ `./sort -f atest`:
```
AA
aa
Aa
aA
Aab
AAb
aab
aAb
b
B
```
Problem here is that for `strcasecmp` "AA" == "Aa". Because of that, output line's position depends on position in input file.
GNU sort does it's own magic. For him "AA" < "Aa" < "aA" and so on. As I figured out it happens because with `-f` option, `sort` folds lower case to upper case, but still thinks that "A" < "a". And what important here, "a" become less than "B". So comparision of ASCII characters will be like that:
```
... < '?' < '@' < 'A' < 'a' < 'B' < 'b'  < ... < 'Z' < 'z' < '[' < '\' < ...
```
Therefore, `strcasecmp` here is useless and `strcmp` with `toupper()` here is useless too (because will be "AA" == "Aa" again). So I guess for `-f` option I will leave my solution (it works somehow).

Reply:
Well, so it seems gnu `sort` preserves the original sorting order among equal strings. Yes, let's keep your solution here.

After your last commit, `compare` looks pretty well! Can you, please, add a small comment in the code explaining that we do that because gnu sort works in the following way with `-f` option: (e.g. Input: `aa AA`, output: `AA aa`)

Reply:
P.S.
An alternative solution could be:
```
    if (ign_case) {
        int ret;
        ret = strcasecmp(fst, snd);
        return ret ? ret : strcmp(fst, snd);
    } else {
```

Reply:
Oh, it looks like your solution is working, didn't thought about that, thanks.
Btw, I've got new problem. Now GNU sort does some magic with `-b` option.
Here's `test` file (`|` character represents beginning of the line):
```
|b
| B
|AA
|B
|aa
| Aa
| b
|    aA
|    B
```
`./sort -b test`:
```
|AA
| Aa
| B
|B
|    B
|    aA
|aa
|b
| b
```
`LC_ALL=C sort -b test`:
```
|AA
| Aa
|    B
| B
|B
|    aA
|aa
| b
|b
```
Thing here is that GNU sort always try to sort. In my implementation "[space]B" == "B" == "[tab]B". Therefore they're placed like in input file. 
For GNU sort "[tab]B" < "[space]B" < "B". So they're sorted like that.
It has logic in that. Even with "equal" lines one still can expect somehow sorted output. Don't know if that implementation was necessary or just for eyecandy. Anyway, it seems like I'll need little more time to solve this.